feat: add sharing info to /workspaces endpoint

geokat opened the pull request:
Similar to https://github.com/coder/coder/pull/19375, this one uses system permissions for fetching actual user and group data.
An implementation that doesn't do that is also buried in the commits, so it should be easy to switch to if needed.

closes: https://github.com/coder/internal/issues/858

geokat committed:
hide `shared_with` in workspace response behind experiment

geokat committed:
codersdk changes

geokat committed:
Deserialize user and group ACLs when reading workspaces from DB

Also fix bug in the deserialization code.

geokat committed:
Fetch user and group data from DB

geokat committed:
Add user and group data to endpoint response

geokat committed:
Only display member and group data if user has required authorization

geokat committed:
Clean up, refactor

geokat committed:
Add integration tests

geokat committed:
Parallelize user/group fetching

geokat committed:
Go back to fetching user and group data with system context

geokat committed:
make gen

Emyrk commented on coderd/database/modelmethods.go:
```diff
@@ -683,9 +683,16 @@ func ConvertWorkspaceRows(rows []GetWorkspacesRow) []Workspace {
+		}
+		if err := workspaces[i].GroupACL.Scan(r.GroupACL); err != nil {
+			return nil, xerrors.Errorf("scan group ACL %q: %w", r.GroupACL, err)
+		}
```
These fields were just missing before?

Another missing field I see is  `OwnerName: r.OwnerName,`

geokat replied:
Done.

However it looks like it isn't used:
https://github.com/coder/coder/blob/geokat/internal-858-sharing-info-in-workspaces-response/coderd/workspaces.go#L2678

But I added it for completeness.

Emyrk reviewed (commented):
Do you think we could do this all in the postgres view?

We have the view 
https://github.com/coder/coder/blob/9e335492709df6d299e06d9d40e9cad99b781d1f/coderd/database/dump.sql#L2911-L2911

I know the ACLs are `jsonb`. If we add this to the view though, that would save us work downstream.

There are api endpoints in this PR that return workspaces without the shared info. I'd rather us be consistent what fields are returned when fetching a workspace.

geokat committed:
Fetch user/group visible info for workspaces ACL actors through workspaces_expanded view

Instead of N+1 in coderd.

geokat committed:
Fix migration numbering: 402 already exists in main

geokat committed:
Add missing OwnerName in ConvertWorkspaceRows

aslilac reviewed (commented):
the go code all looks largely fine but I'd love to have a second pair of eyes on the sql `view` because it's quite involved.

geokat commented on coderd/workspaces.go:
```diff
@@ -2662,9 +2706,62 @@ func convertWorkspace(
+		// TODO(geokat): using a pointer that's serialized with
+		// "omitempty" so that we can hide it behind an experiment. This
+		// won't be necessary once workspace sharing is in GA.
+		SharedWith: sharedWith,
```
I agree it's a bit hacky, but please see my explanation below.

geokat replied:
Pushed an update that gets rid of the pointer usage.

geokat commented on codersdk/workspaces.go:
```diff
@@ -73,7 +73,8 @@ type Workspace struct {
 	// TaskID, if set, indicates that the workspace is relevant to the given codersdk.Task.
-	TaskID uuid.NullUUID `json:"task_id,omitempty"`
+	TaskID     uuid.NullUUID           `json:"task_id,omitempty"`
+	SharedWith *[]SharedWorkspaceActor `json:"shared_with,omitempty"`
```
I was going for the "include empty array in the response json if workspace is not shared, but only if the experiment is enabled" effect.  The use of a pointer, although hacky, enabled this.

Having said this, I don't know what the convention is for empty arrays (to `omitempty` or not).  For example, empty [`latest_build.resources.agents.apps.statuses`](https://github.com/coder/coder/blob/geokat/internal-858-sharing-info-in-workspaces-response/codersdk/workspaceapps.go#L97) serialize to `[]` and so do [`agents.subsystems`](https://github.com/coder/coder/blob/geokat/internal-858-sharing-info-in-workspaces-response/codersdk/workspaceagents.go#L169), but then again there are [these](
https://github.com/coder/coder/blob/geokat/internal-858-sharing-info-in-workspaces-response/codersdk/workspacebuilds.go#L107).

You could also say that `shared_with: []` is semantically different from "no info at all", but that depends on the conventions?

But please let me know if you still prefer a slice with `omitempty`; it should be an easy change.

geokat replied:
After a chat with @Emyrk, the takeaway is that it's OK to omit empty `shared_with` from the response now, so I will push an update that gets rid of the pointer usage.

Emyrk commented on coderd/database/migrations/000403_workspaces_expanded_acl_actor_info.up.sql:
```diff
@@ -0,0 +1,63 @@
+            )
+            FROM jsonb_each(workspaces.user_acl) AS acl
+            LEFT JOIN visible_users vu ON vu.id = acl.key::uuid
+        ), '{}'::jsonb) AS user_acl,
```
Upon further checking, the actual query time without analyze is exactly the same. So scratch this

geokat committed:
Get rid of SharedWith pointer usage which enabled support for `shared_with: []` in response if experiment was enabled

geokat committed:
chore: fix test helper function so that it doesn't save nulls for user/group acl in workspace test table

Because such nulls break tests fetching the new workspace_expanded
view, with the following error:

`cannot call jsonb_each on a non-object`

(null is a non-object, that's why we use `{}`as default).

This begs the question if we want to adopt check constraints for JSONB
columns across our schema, e.g.:

`ALTER TABLE workspaces
   ADD CONSTRAINT group_acl_is_object CHECK (jsonb_typeof(group_acl) = 'object');`

aslilac reviewed (changes requested):
sorry, noticed something on a second look that I think we should address

Emyrk commented on coderd/database/dbgen/dbgen.go:
```diff
@@ -440,10 +440,18 @@ func Workspace(t testing.TB, db database.Store, orig database.WorkspaceTable) da
+		groupACL := orig.GroupACL
+		if groupACL == nil {
+			groupACL = database.WorkspaceACL{}
+		}
```
Ah, do we need a db migration and constraint to fix it? 

That would be ok

geokat committed:
refactor: use separate columns in the workspaces_expanded view for the acl actors display info

geokat commented on codersdk/workspaces.go:
```diff
@@ -695,6 +696,14 @@ type WorkspaceUser struct {
 	Role WorkspaceRole `json:"role" enums:"admin,use"`
 }
 
+type SharedWorkspaceActor struct {
```
Good point, thanks! Pushed an update that reads the ACL actors' display info as separate columns on the `workspaces_expanded` view.

The query behind this version of the view is slightly (up to 20%) faster, likely because it doesn't need to parse saved ACL JSON:

[old version](https://explain.dalibo.com/plan/a89d65467b0bh3f3)
[new version](https://explain.dalibo.com/plan/4d99b346dg91bd31)

The downside is more data on the wire and extra boilerplate.

aslilac reviewed (approved):
thanks!

Emyrk reviewed (approved):
The split into 2 columns is fine. Not ideal, but if it's more efficient, than that is a win.

geokat commented on coderd/workspaces.go:
```diff
@@ -2662,9 +2704,59 @@ func convertWorkspace(
+			ActorType: codersdk.SharedWorkspaceActorTypeUser,
+			Roles:     []codersdk.WorkspaceRole{convertToWorkspaceRole(aclEntry.Permissions)},
+			Name:      workspace.UserACLDisplayInfo[id].Name,
+			AvatarURL: workspace.UserACLDisplayInfo[id].AvatarURL,
```
It's a map, so no dereferencing here. Nil is the zero value, so it's safe:

https://go.dev/play/p/DNA42hnXVH3

I could add a check to be defensive in case this changes to a pointer to a map in the future, but not sure if it's necessary?

Emyrk replied:
TIL you can deref a nil map entry :+1:

geokat committed:
Added some comments
